dodiscover/ci/categorical_test.py

```diff
@@ -0,0 +1,200 @@
+
+
+# copied from pgmpy
+def power_divergence(X, Y, Z, data, lambda_="cressie-read"):
+    """
+    Computes the Cressie-Read power divergence statistic [1]. The null hypothesis
+    for the test is X is independent of Y given Z. A lot of the frequency comparison
+    based statistics (eg. chi-square, G-test etc) belong to power divergence family,
+    and are special cases of this test.
+
+    Parameters
+    ----------
+    X: int, string, hashable object
+        A variable name contained in the data set
+
+    Y: int, string, hashable object
+        A variable name contained in the data set, different from X
+
+    Z: list, array-like
+        A list of variable names contained in the data set, different from X and Y.
+        This is the separating set that (potentially) makes X and Y independent.
+        Default: []
+
+    data: pandas.DataFrame
+        The dataset on which to test the independence condition.
+
+    lambda_: float or string
+        The lambda parameter for the power_divergence statistic. Some values of
+        lambda_ results in other well known tests:
+            "pearson"             1          "Chi-squared test"
+            "log-likelihood"      0          "G-test or log-likelihood"
+            "freeman-tukey"     -1/2        "freeman-tukey Statistic"
+            "mod-log-likelihood"  -1         "Modified Log-likelihood"
+            "neyman"              -2         "Neyman's statistic"
+            "cressie-read"        2/3        "The value recommended in the paper
+                                             :footcite:`cressieread1984`"
+
+    Returns
+    -------
+    CI Test Results: tuple
+        Returns a tuple (chi, p_value, dof). `chi` is the
+        chi-squared test statistic. The `p_value` for the test, i.e. the
+        probability of observing the computed chi-square statistic (or an even
+        higher value), given the null hypothesis that X \u27C2 Y | Zs is True.
+        If boolean = True, returns True if the p_value of the test is greater
+        than `significance_level` else returns False.
+
+    See Also
+    --------
+    scipy.stats.power_divergence
+
+    References
+    ----------
+    .. footbibliography::
+
+    Examples
+    --------
+    >>> import pandas as pd
+    >>> import numpy as np
+    >>> data = pd.DataFrame(np.random.randint(0, 2, size=(50000, 4)), columns=list('ABCD'))
+    >>> data['E'] = data['A'] + data['B'] + data['C']
+    >>> chi_square(X='A', Y='C', Z=[], data=data, boolean=True, significance_level=0.05)
+    True
+    >>> chi_square(X='A', Y='B', Z=['D'], data=data, boolean=True, significance_level=0.05)
+    True
+    >>> chi_square(X='A', Y='B', Z=['D', 'E'], data=data, boolean=True, significance_level=0.05)
+    False
+    """
+
+    # Step 1: Check if the arguments are valid and type conversions.
+    if hasattr(Z, "__iter__"):
+        Z = list(Z)
+    else:
+        raise (f"Z must be an iterable. Got object type: {type(Z)}")
```

Comment:
What type of exception? I guess ```raise ValueError(...)```?

Another remark: I don't think it necessary to test for these kind of things. Just add an appropriate typing to the variables and then assume that the correct data type is given. Otherwise, you will need to check all kind of things, which is not really worth it.

Reply:
Okay that works too. Modified